File: client-apps/cli/cmd/stigmer/root/seedpack.go
package root

import (
	"os"

	"github.com/rs/zerolog"
	"github.com/spf13/cobra"
	"github.com/stigmer/stigmer/client-apps/cli/internal/cli/clierr"
	"github.com/stigmer/stigmer/client-apps/cli/internal/cli/config"
	"github.com/stigmer/stigmer/client-apps/cli/internal/cli/seedpackbootstrap"
	"github.com/stigmer/stigmer/client-apps/cli/pkg/clioutput"
)

// NewSeedpackCommand creates the seedpack command group.
func NewSeedpackCommand() *cobra.Command {
	cmd := &cobra.Command{
		Use:   "seedpack",
		Short: "Manage system seedpack resources",
		Long: `Manage the embedded system seedpack — the built-in agents, skills,
and MCP server definitions that ship with the CLI.

The seedpack contains system resources (agent-creator, skill-creator,
mcp-server-creator, assistant) under the "stigmer" organization.

In local mode, the seedpack is applied automatically on server startup.
In cloud mode, use 'stigmer seedpack apply' to bootstrap the cloud
backend with these system resources.`,
	}

	cmd.AddCommand(newSeedpackApplyCommand())
	cmd.AddCommand(newSeedpackStatusCommand())

	return cmd
}

func newSeedpackApplyCommand() *cobra.Command {
	var force bool
	var jsonOutput, quietOutput bool

	cmd := &cobra.Command{
		Use:   "apply",
		Short: "Apply seedpack to the configured backend",
		Long: `Extract the embedded seedpack and apply it to whatever backend the
CLI is currently configured for (local or cloud).

The apply is idempotent: if the seedpack content has not changed since
the last successful apply, the operation is skipped. Use --force to
re-apply unconditionally.

The target organization defaults to "stigmer" and can be overridden
via the STIGMER_SEEDPACK_ORG environment variable.`,
		Example: `  # Apply seedpack to the current backend
  stigmer seedpack apply

  # Force re-apply even if already up to date
  stigmer seedpack apply --force

  # Apply to a cloud backend
  stigmer config backend set cloud
  stigmer auth login
  stigmer seedpack apply`,
		Run: func(cmd *cobra.Command, args []string) {
			handleSeedpackApply(force, resolveResultFormat(jsonOutput, quietOutput))
		},
	}

	cmd.Flags().BoolVar(&force, "force", false, "re-apply even if the seedpack has not changed")
	addResultFormatFlags(cmd, &jsonOutput, &quietOutput)

	return cmd
}

func handleSeedpackApply(force bool, format clioutput.OutputFormat) {
	renderer := clioutput.NewRenderer(format, os.Stdout, os.Stderr)

	markerDir, err := resolveMarkerDir()
	if err != nil {
		clierr.Handle(err)
		return
	}

	err = seedpackbootstrap.Apply(seedpackbootstrap.Options{
		MarkerDir: markerDir,
		Force:     force,
		Verbose:   zerolog.GlobalLevel() <= zerolog.DebugLevel,
	})
	if err != nil {
		clierr.Handle(err)
		return
	}

	result := clioutput.Success("Seedpack applied to %s backend", resolveBackendLabel())
	renderer.Render(result)
}

func newSeedpackStatusCommand() *cobra.Command {
	var jsonOutput, quietOutput bool

	cmd := &cobra.Command{
		Use:   "status",
		Short: "Show seedpack bootstrap status",
		Long:  `Check whether the embedded seedpack has been applied to the current backend.`,
		Run: func(cmd *cobra.Command, args []string) {
			handleSeedpackStatus(resolveResultFormat(jsonOutput, quietOutput))
		},
	}

	addResultFormatFlags(cmd, &jsonOutput, &quietOutput)
	return cmd
}

func handleSeedpackStatus(format clioutput.OutputFormat) {
	renderer := clioutput.NewRenderer(format, os.Stdout, os.Stderr)

	hash, err := seedpackbootstrap.CurrentHash()
	if err != nil {
		clierr.Handle(err)
		return
	}

	markerDir, err := resolveMarkerDir()
	if err != nil {
		clierr.Handle(err)
		return
	}

	applied, storedHash := seedpackbootstrap.MarkerStatus(markerDir)

	result := clioutput.Success("Seedpack status")
	sec := result.AddSection("")
	sec.Field("Backend", resolveBackendLabel())
	sec.Field("Embedded Hash", hash)

	if applied {
		sec.Field("Applied Hash", storedHash)
		if storedHash == hash {
			sec.Field("Status", "Up to date")
		} else {
			sec.Field("Status", "Outdated (run 'stigmer seedpack apply' to update)")
		}
	} else {
		sec.Field("Status", "Not applied (run 'stigmer seedpack apply')")
	}

	renderer.Render(result)
}

// resolveMarkerDir returns the directory for the seedpack idempotency marker,
// based on the current backend type.
func resolveMarkerDir() (string, error) {
	cfg, err := config.Load()
	if err != nil {
		return "", err
	}

	configDir, err := config.GetConfigDir()
	if err != nil {
		return "", err
	}

	switch cfg.Backend.Type {
	case config.BackendTypeLocal:
		dataDir, err := config.GetDataDir()
		if err != nil {
			return "", err
		}
		return dataDir, nil
	default:
		return configDir, nil
	}
}

// resolveBackendLabel returns a human-readable label for the current backend.
func resolveBackendLabel() string {
	cfg, err := config.Load()
	if err != nil {
		return "unknown"
	}

	switch cfg.Backend.Type {
	case config.BackendTypeCloud:
		endpoint := "api.stigmer.ai:443"
		if cfg.Backend.Cloud != nil && cfg.Backend.Cloud.Endpoint != "" {
			endpoint = cfg.Backend.Cloud.Endpoint
		}
		return "cloud (" + endpoint + ")"
	default:
		return "local"
	}
}

File: client-apps/cli/src/commands/seedpack.ts
// `stigmer seedpack apply|status` — manage the embedded system seedpack.
//
// The seedpack (system agents, skills, MCP servers, workflows under the
// "stigmer" org) is applied automatically on `stigmer up` in local mode; these
// commands bootstrap it explicitly against whatever backend is configured
// (local or cloud). Apply is idempotent via a content-hash marker.
//
// Heavy modules load lazily inside the action so `--help` stays fast (DD-001).

import type { Command } from "commander";
import { configDir, dataDir, isCloudMode, load } from "../config/index.js";
import { CommandResult, type OutputFlags, renderResult } from "../output/index.js";
import { addResultFlags, resultFormat } from "./shared.js";

interface ApplyFlags extends OutputFlags {
  force?: boolean;
}

export function registerSeedpack(program: Command): void {
  const seedpack = program.command("seedpack").description("manage the embedded system seedpack");

  const apply = seedpack
    .command("apply")
    .description("apply the seedpack to the configured backend")
    .option("--force", "re-apply even if the seedpack has not changed", false)
    .action((options: ApplyFlags) => runApply(options));
  addResultFlags(apply);

  const status = seedpack
    .command("status")
    .description("show seedpack bootstrap status")
    .action((options: OutputFlags) => runStatus(options));
  addResultFlags(status);
}

async function runApply(flags: ApplyFlags): Promise<void> {
  const [{ connectBackend }, { ensureAuthenticated }, { applySeedpack }] = await Promise.all([
    import("../backend.js"),
    import("../config/index.js"),
    import("../local/seedpack/apply.js"),
  ]);

  const client = connectBackend();
  ensureAuthenticated(client.config);

  const result = await applySeedpack(
    {
      controller: client.controller,
      stigmer: client.stigmer,
      info: (line) => process.stderr.write(`${line}\n`),
      warn: (line) => process.stderr.write(`${line}\n`),
    },
    { markerDir: markerDir(), force: flags.force === true },
  );

  const label = backendLabel();
  const message = result.applied
    ? `Seedpack applied to ${label} backend`
    : `Seedpack already up to date on ${label} backend (use --force to re-apply)`;
  renderResult(CommandResult.success(message), resultFormat(flags));
}

async function runStatus(flags: OutputFlags): Promise<void> {
  const { seedpackContentHash } = await import("../local/seedpack/apply.js");
  const { readMarker } = await import("../local/seedpack/content.js");

  const hash = seedpackContentHash();
  const stored = readMarker(markerDir());

  const result = CommandResult.success("Seedpack status");
  const section = result.addSection("");
  section.field("Backend", backendLabel());
  section.field("Embedded Hash", hash);
  if (stored === null) {
    section.field("Status", "Not applied (run 'stigmer seedpack apply')");
  } else {
    section.field("Applied Hash", stored);
    section.field("Status", stored === hash ? "Up to date" : "Outdated (run 'stigmer seedpack apply' to update)");
  }
  renderResult(result, resultFormat(flags));
}

// Marker lives with the backend's state: the data dir for local, the config dir
// for cloud (mirrors the Go CLI's resolveMarkerDir).
function markerDir(): string {
  return isCloudMode(load()) ? configDir() : dataDir();
}

function backendLabel(): string {
  const config = load();
  if (!isCloudMode(config)) return "local";
  const endpoint = config.backend.cloud?.endpoint ?? "api.stigmer.ai:443";
  return `cloud (${endpoint})`;
}
